### PR TITLE
[bugfix][transactions] Make TxnTransitMetadata.topicPartitions immutable

### PR DESCRIPTION
**Pull request by Demogorgon314**
### Motivation

https://github.com/apache/kafka/blob/trunk/core/src/main/scala/kafka/coordinator/transaction/TransactionMetadata.scala#L190

The `TxnTransitMetadata.topicPartitions` should be immutable, otherwise it will cause race conditions.

### Modifications

Make TxnTransitMetadata.topicPartitions immutable

### Documentation

Check the box below.

Need to update docs? 

- [ ] `doc-required` 
  
  (If you need help on updating docs, create a doc issue)
  
- [x] `no-need-doc` 
  
  (Please explain why)
  
- [ ] `doc` 
  
  (If this PR contains doc changes)

### CODE REVIEW

**Review by eolivelli — DISMISSED** [review]
+1

File: kafka-impl/src/main/java/io/streamnative/pulsar/handlers/kop/coordinator/transaction/TransactionMetadata.java [hunk 1]
```diff
@@ -434,14 +437,14 @@ public TxnTransitMetadata prepareAddPartitions(Set<TopicPartition> addedTopicPar
         if (topicPartitions != null) {
             newPartitionSet.addAll(topicPartitions);
         }
-        newPartitionSet.addAll(new HashSet<>(addedTopicPartitions));
+        newPartitionSet.addAll(addedTopicPartitions);
         return prepareTransitionTo(TransactionState.ONGOING, producerId, producerEpoch, lastProducerEpoch,
-                txnTimeoutMs, newPartitionSet, newTxnStartTimestamp, updateTimestamp);
+                txnTimeoutMs, ImmutableSet.copyOf(newPartitionSet), newTxnStartTimestamp, updateTimestamp);
```
Comment (BewareMyPower) [inline kafka-impl/src/main/java/io/streamnative/pulsar/handlers/kop/coordinator/transaction/TransactionMetadata.java:442]: Here we don't need to copy again. You can just use `ImmutableSet.copyOf((topicPartitions != null) ? topicPartitions : ImmutableSet.of());`

Comment (Demogorgon314) [reply kafka-impl/src/main/java/io/streamnative/pulsar/handlers/kop/coordinator/transaction/TransactionMetadata.java:442]: Here we also added all `addedTopicPartitions ` into `newPartitionSet`.

Comment (BewareMyPower) [reply kafka-impl/src/main/java/io/streamnative/pulsar/handlers/kop/coordinator/transaction/TransactionMetadata.java:442]: Oh I missed it. Then we can replace it with:

```java
        ImmutableSet<TopicPartition> partitions = ImmutableSet.<TopicPartition>builder()
                .addAll((topicPartitions != null) ? topicPartitions : ImmutableSet.of())
                .addAll(addedTopicPartitions)
                .build();
```

Comment (Demogorgon314) [reply kafka-impl/src/main/java/io/streamnative/pulsar/handlers/kop/coordinator/transaction/TransactionMetadata.java]: Fixed.

**Review by BewareMyPower — APPROVED** [review]

**Review by gaoran10 — APPROVED** [review]

**Pull request merged**